src/Lib.P2P.Tests/AutoDialerTest.cs

```diff
@@ -0,0 +1,203 @@
+        {
+            var swarmA = new SwarmService {LocalPeer = peerA};
+            await swarmA.StartAsync();
+            await swarmA.StartListeningAsync("/ip4/127.0.0.1/tcp/0");
+
+            var swarmB = new SwarmService {LocalPeer = peerB};
+            await swarmB.StartAsync();
+            var peerBAddress = await swarmB.StartListeningAsync("/ip4/127.0.0.1/tcp/0");
+
+            try
+            {
+                using (new AutoDialer(swarmA) {MinConnections = 0})
+                {
+                    var other = swarmA.RegisterPeerAddress(peerBAddress);
+
+                    // wait for the connection.
+                    var endTime = DateTime.Now.AddSeconds(3);
+                    while (other.ConnectedAddress == null)
+                    {
+                        if (DateTime.Now > endTime)
+                        {
+                            return;
+                        }
+                        
+                        await Task.Delay(100);
+                    }
+
+                    Assert.Fail("Autodial should not happen");
+                }
+            }
+            finally
+            {
+                await swarmA.StopAsync();
+                await swarmB.StopAsync();
+            }
+        }
+
+        [TestMethod]
+        public async Task Connects_OnPeerDisconnected_When_Below_MinConnections()
+        {
+            var swarmA = new SwarmService {LocalPeer = peerA};
+            await swarmA.StartAsync();
+            await swarmA.StartListeningAsync("/ip4/127.0.0.1/tcp/0");
+
+            var swarmB = new SwarmService {LocalPeer = peerB};
+            await swarmB.StartAsync();
+            var peerBAddress = await swarmB.StartListeningAsync("/ip4/127.0.0.1/tcp/0");
+
+            var swarmC = new SwarmService {LocalPeer = peerC};
+            await swarmC.StartAsync();
+            var peerCAddress = await swarmC.StartListeningAsync("/ip4/127.0.0.1/tcp/0");
+
+            var isBConnected = false;
+            swarmA.ConnectionEstablished += (s, conn) =>
+            {
+                if (conn.RemotePeer == peerB)
+                {
+                    isBConnected = true;
+                }
+            };
+
+            try
+            {
+                using (new AutoDialer(swarmA) {MinConnections = 1})
+                {
+                    swarmA.RegisterPeerAddress(peerBAddress);
+                    var c = swarmA.RegisterPeerAddress(peerCAddress);
+
+                    // wait for the peer B connection.
+                    var endTime = DateTime.Now.AddSeconds(3);
+                    while (!isBConnected)
+                    {
+                        if (DateTime.Now > endTime)
+                        {
+                            Assert.Fail("Did not do autodial on peer discovered");
+                        }
+                        
+                        await Task.Delay(100); // get cancellaton token
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)